### PR TITLE
Fix #14577: always include every new line ident when diffing hypothesis lines

### LINKED ISSUE #14577: revert vs. hypotheses grouping
(this is broken out from https://github.com/coq/coq/issues/14546)

This is in 8.13.2, with PG 4.5-git (I have not checked coqide or other UIs).
Hypotheses grouping does not occur following the `revert` tactic.  Any subsequent nontrivial (not idtac) tactic will fix grouping if it is not part of the same Ltac `;`-separated command.

```Coq
Goal nat -> nat -> bool -> nat -> nat -> True.
  intros i j b k l. revert b. cbn.
```
The sequence of goal contexts following each tactic is:
```
1 subgoal (ID 7)
  i, j : nat
  b : bool
  k, l : nat
  ============================
  True

1 subgoal (ID 9)
  i, j : nat
  k, l : nat
  ============================
  bool -> True

1 subgoal (ID 10)
  i, j, k, l : nat
  ============================
  bool -> True
```
Note that the nats are grouped in two separate but adjacent lines following the `revert`, and that the following `cbn` collapses them to a single line.  Note also that composing tactics to follow `revert` with `;` in the same Ltac command (such as `revert;cbn`) does not result in a single line.  This means that Ltac definitions containing `revert` will not immediately result in the most collapsed grouping of hypotheses.

### PR DESCRIPTION
**Pull request by jfehrle**
~~Always Include every new line ident when diffing hypotheses lines so it looks similar to the undiffed hypotheses.~~

~~Spent an hour or two trying to add a test for this but couldn't get the test to work.  That's difficult to debug when print statements added in the Coq code for debugging don't show anything on the console or in the test log--there used to be a way to do that.  :-(~~

After looking at more test cases, I decided to rewrite the algorithm completely based on some better ideas.  The behavior of the new algorithm is easy to describe and the new code is fairly easy to follow.  I can't say that about the original algorithm.  The description is just before `Proof_diffs.diff_hyps`.

Fixes / closes #14577

### CODE REVIEW

File: printing/proof_diffs.ml [hunk 1]
```diff
@@ -192,15 +193,14 @@ let diff_hyps o_line_idents o_map n_line_idents n_map =
...
               | (Some fst, Some this) when fst == this ->  (* yes, == *)
-                add n_ids_list ident3 n_map;
                 (* include, in old order, all undone Removed idents in old *)
                 List.iter (fun x -> if x = ident3 || not (is_done x o_map) && not (exists x n_map) then
                                     (add o_ids_list x o_map)) fst
-              | (_, None) ->
-                add n_ids_list ident3 n_map (* include all undone Added idents in new *)
               | _ -> ()
+              end;
+              add n_ids_list ident3 n_map
```
Comment (herbelin) [inline printing/proof_diffs.ml:203]: Is this the key swap of line of the PR?

**Review by herbelin — APPROVED** [review]
As a side meta-comment, I may suggest a different commit for the fix and for the cosmetic changes.

**Comment by jfehrle** [comment]
Yup.  Also need to fix the test, which seems to pass when I run it locally :-(

File: printing/proof_diffs.ml (hunk 1 above)
Comment (jfehrle) [reply printing/proof_diffs.ml:203]: Yes.

**Comment by jfehrle** [comment]
@herbelin Note that I updated the PR description above.  Please look at the new algorithm (in `Proof_diffs.diff_hyps`).  If possible, it would be great to get this into 8.15.  Let me know if you have questions/comments.  Thanks.

The ci-gappa failure looks unrelated.

**Comment by SkySkimmer** [comment]
@herbelin are you assigning?

**Comment by herbelin** [comment]
>  The behavior of the new algorithm is easy to describe and the new code is failyr easy to follow. I can't say that about the original algorithm. 

I had a quick look and it looks great.

I'm assigning and merging soon.

**Comment by herbelin** [comment]
@coqbot: merge now

**Comment by jfehrle** [comment]
Thanks Hugo!

**Pull request merged**